### ISSUE TITLE
A question on the function of class Model in loss.py

### ISSUE THREAD

**Issue by yaoyiran** [drboog/Lafite#23]
Hi, when I read the code I noticed that you mentioned "# This implementation is incorrect, it should be sim=sim/temp." for [L149](https://github.com/drboog/Lafite/blob/main/training/loss.py#L149) in loss.py. I later found that in [L45](https://github.com/drboog/Lafite/blob/main/training/loss.py#L45), there is the same "bug". If I correct L149, do I also need to correct L45?

However, when I read the code I found that the Model class in loss.py is only defined in L75-L77 in loss.py but not called later. May I know what is the function of Model class here?

**Comment by drboog** [comment]
1. As I mentioned in the paper, although " sim=sim/temp" is the correct implementation for vanilla contrastive loss, but using an exponential operation here will easily reproduce our results in the paper, it's still a revised version of contrastive loss, just with a different penalty.

2. Simply search "mapper", you can see line 187- 194 at loss.py, that's where the Model is called. We proposed two perturbation in the paper, Gaussian and learnable, the Model is for the second one.

**Comment by yaoyiran** [comment]
Thank you so much for your message! I also noticed that "exponential sharpening" is added in the supplementary material. That's great!

May I have two more technical questions I encountered when running your code?

(1) In [L178-L185](https://github.com/drboog/Lafite/blob/main/training/loss.py#L178-L185) in loss.py, gaussian perturbation is applied; this corresponds to the paper's descriptions on language-free settings. I was wondering in "Training with ground-truth pairs", i.e., **fully supervised settings**, is the gaussian noise still also kept as a way of data augmentation?

(2) I run the dataset_tool.py several times on the original COCO data, and found that the preprocessed [MS-COCO Training Set](https://drive.google.com/file/d/1b82BCh65XxwR-TiA8zu__wwiEHLCgrw2/view?usp=sharing) and [MS-COCO Validation Set](https://drive.google.com/file/d/1qBy5rPfo1go4d-PjF_Gu0kESCZ9Nt1Ta/view?usp=sharing) released on the LAFITE's GitHub page is actually derived without cropping, i.e., transform should be None (by not specifying it in the command line). Could you verify this? (If you still remember the command line you used with dataset_tool.py for **fully supervised setting** and could share it, it will be very helpful.)

**Comment by drboog** [comment]
1. You can see in the code Gaussian perturbation is still used in fully-supervised setting, with a small perturbation level 0.1. Yes, it is a data augmentation.

2. I don't remember, but it's basically a question of using center-crop or not. So compare a image from pre-processed zip file to corresponding original image from COCO website should tell you whether center-crop is used. The idea behind center-crop is that, images of a given dataset is not always square, directly resize them into 256*256 may lead to distortion. Training a model on many distorted images may lead to undesirable generations. If the images are all roughly-square, then using center-crop or not do not have much difference.

**Comment by yaoyiran** [comment]
Thanks! It seems true that center-crop was not used. Here is an example, comparing a COCO original image with an image from [MS-COCO Training Set](https://drive.google.com/file/d/1b82BCh65XxwR-TiA8zu__wwiEHLCgrw2/view?usp=sharing). I think center-crop has its pros and cons: distortion vs loss of content.

![COCO_train2014_000000001505](https://user-images.githubusercontent.com/25559589/195433366-09fea48f-da0d-46e7-96c1-0bad9cab9398.jpg)
![img00000206](https://user-images.githubusercontent.com/25559589/195433390-5a5184cb-0dd5-454c-81e5-83e88357c367.png)
